build fails

linux/LinuxProcessList.c:632:13: warning: In the GNU C Library, "minor" is defined
 by <sys/sysmacros.h>. For historical compatibility, it is
 currently defined by <sys/types.h> as well, but we plan to
 remove this soon. To use "minor", include <sys/sysmacros.h>
 directly. If you did not intend to use a system-defined macro
 "minor", you should undefine it after including <sys/types.h>.
make[1]: *** [Makefile:1731: linux/htop-LinuxProcessList.o] Error 1
make[1]: Leaving directory '/usr/local/src/tools/htop/trunk'
make: *** [Makefile:521: all] Error 2

Well, now it definitely fails on GCC-7.2.0/Clang-5.0.0 (maybe, on 4 too) and glibc-2.26:
```make
>>> Compiling source in /var/tmp/portage/sys-process/htop-9999/work/htop-9999 ...
make -j5 -s -l2
In file included from Object.h:13:0,
                 from Process.h:24,
                 from Process.c:8:
Process.c: In function ‘Process_writeField’:
Process.c:463:50: warning: implicit declaration of function ‘major’ [-Wimplicit-function-declaration]
    case TTY_NR: xSnprintf(buffer, n, "%3u:%3u ", major(this->tty_nr), minor(this->tty_nr)); break;
                                                  ^
XAlloc.h:20:76: note: in definition of macro ‘xSnprintf’
 #define xSnprintf(fmt, len, ...) do { int _l=len; int _n=snprintf(fmt, _l, __VA_ARGS__); if (!(_n > -1 && _n < _l)) { curs_set(1); endwin(); err(1, NULL); } } while(0)
                                                                            ^~~~~~~~~~~
Process.c:463:71: warning: implicit declaration of function ‘minor’; did you mean ‘mknod’? [-Wimplicit-function-declaration]
    case TTY_NR: xSnprintf(buffer, n, "%3u:%3u ", major(this->tty_nr), minor(this->tty_nr)); break;
                                                                       ^
XAlloc.h:20:76: note: in definition of macro ‘xSnprintf’
 #define xSnprintf(fmt, len, ...) do { int _l=len; int _n=snprintf(fmt, _l, __VA_ARGS__); if (!(_n > -1 && _n < _l)) { curs_set(1); endwin(); err(1, NULL); } } while(0)
                                                                            ^~~~~~~~~~~
linux/LinuxProcessList.c: In function ‘LinuxProcessList_updateTtyDevice’:
linux/LinuxProcessList.c:712:23: warning: implicit declaration of function ‘major’ [-Wimplicit-function-declaration]
    unsigned int maj = major(tty_nr);
                       ^~~~~
linux/LinuxProcessList.c:713:23: warning: implicit declaration of function ‘minor’; did you mean ‘mknod’? [-Wimplicit-function-declaration]
    unsigned int min = minor(tty_nr);
                       ^~~~~
                       mknod
linux/LinuxProcessList.c:736:47: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
          if (err == 0 && major(sstat.st_rdev) == maj && minor(sstat.st_rdev) == min) return fullPath;
                                               ^~
linux/LinuxProcessList.c:736:78: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
          if (err == 0 && major(sstat.st_rdev) == maj && minor(sstat.st_rdev) == min) return fullPath;
                                                                              ^~
linux/LinuxProcessList.c:740:47: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
          if (err == 0 && major(sstat.st_rdev) == maj && minor(sstat.st_rdev) == min) return fullPath;
                                               ^~
linux/LinuxProcessList.c:740:78: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
          if (err == 0 && major(sstat.st_rdev) == maj && minor(sstat.st_rdev) == min) return fullPath;
                                                                              ^~
linux/LinuxProcessList.c:734:10: warning: ignoring return value of ‘asprintf’, declared with attribute warn_unused_result [-Wunused-result]
          asprintf(&fullPath, "%s/%d", ttyDrivers[i].path, idx);
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
linux/LinuxProcessList.c:738:10: warning: ignoring return value of ‘asprintf’, declared with attribute warn_unused_result [-Wunused-result]
          asprintf(&fullPath, "%s%d", ttyDrivers[i].path, idx);
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
linux/LinuxProcessList.c:749:4: warning: ignoring return value of ‘asprintf’, declared with attribute warn_unused_result [-Wunused-result]
    asprintf(&out, "/dev/%u:%u", maj, min);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
htop-Process.o: In function `Process_writeField':
Process.c:(.text+0xe8b): undefined reference to `minor'
Process.c:(.text+0xe98): undefined reference to `major'
linux/htop-LinuxProcessList.o: In function `LinuxProcessList_recurseProcTree':
LinuxProcessList.c:(.text+0x1075): undefined reference to `major'
LinuxProcessList.c:(.text+0x1082): undefined reference to `minor'
LinuxProcessList.c:(.text+0x113c): undefined reference to `major'
LinuxProcessList.c:(.text+0x11a5): undefined reference to `major'
LinuxProcessList.c:(.text+0x11bd): undefined reference to `minor'
LinuxProcessList.c:(.text+0x17db): undefined reference to `minor'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:733: htop] Error 1
make: *** [Makefile:528: all] Error 2
```

This is how it looks here, unless I include sysmacros.h:
```
make[1]: Entering directory '/var/portage/tmp/portage/sys-process/htop-2.1.0/work/htop-2.1.0'
x86_64-pc-linux-gnu-gcc -pedantic -Wall -Wextra -std=c99 -D_XOPEN_SOURCE_EXTENDED -DSYSCONFDIR=\"/etc\" -I"./linux" -rdynamic  -O3 -pipe -march=native    -o htop   htop-AvailableMetersPanel.o htop-CategoriesPanel.o htop-CheckItem.o htop-ClockMeter.o htop-ColorsPanel.o htop-ColumnsPanel.o htop-CPUMeter.o htop-CRT.o htop-MainPanel.o htop-DisplayOptionsPanel.o htop-FunctionBar.o htop-Hashtable.o htop-Header.o htop-htop.o htop-ListItem.o htop-LoadAverageMeter.o htop-MemoryMeter.o htop-Meter.o htop-MetersPanel.o htop-Object.o htop-Panel.o htop-BatteryMeter.o htop-Process.o htop-ProcessList.o htop-RichString.o htop-ScreenManager.o htop-Settings.o htop-SignalsPanel.o htop-StringUtils.o htop-SwapMeter.o htop-TasksMeter.o htop-UptimeMeter.o htop-TraceScreen.o htop-UsersTable.o htop-Vector.o htop-AvailableColumnsPanel.o htop-AffinityPanel.o htop-HostnameMeter.o htop-OpenFilesScreen.o htop-Affinity.o htop-IncSet.o htop-Action.o htop-EnvScreen.o htop-InfoScreen.o htop-XAlloc.o linux/htop-Platform.o linux/htop-IOPriorityPanel.o linux/htop-IOPriority.o linux/htop-LinuxProcess.o linux/htop-LinuxProcessList.o linux/htop-LinuxCRT.o linux/htop-Battery.o  -L/usr/lib64 -lncursesw -lm  
htop-Process.o: In function `Process_writeField':
Process.c:(.text+0xe4b): undefined reference to `minor'
Process.c:(.text+0xe58): undefined reference to `major'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:733: htop] Error 1
make[1]: Leaving directory '/var/portage/tmp/portage/sys-process/htop-2.1.0/work/htop-2.1.0'
make: *** [Makefile:528: all] Error 2
```

--- a/Process.c
+++ b/Process.c
@@ -28,6 +28,12 @@ in the source distribution for its full text.
 #include <time.h>
 #include <assert.h>
 #include <math.h>
+#ifdef MAJOR_IN_MKDEV
+#include <sys/mkdev.h>
+#elif defined(MAJOR_IN_SYSMACROS) || \
+   (defined(HAVE_SYS_SYSMACROS_H) && HAVE_SYS_SYSMACROS_H)
+#include <sys/sysmacros.h>
+#endif
 
 #ifdef __ANDROID__
 #define SYS_ioprio_get __NR_ioprio_get

--- a/configure.ac
+++ b/configure.ac
@@ -61,6 +61,16 @@ AC_CHECK_HEADERS([stdlib.h string.h strings.h sys/param.h sys/time.h unistd.h],[
 ])
 AC_CHECK_HEADERS([execinfo.h],[:],[:])
 
+AC_HEADER_MAJOR
+dnl glibc 2.25 deprecates 'major' and 'minor' in <sys/types.h> and requires to
+dnl include <sys/sysmacros.h>. However the logic in AC_HEADER_MAJOR has not yet
+dnl been updated in Autoconf 2.69, so use a workaround:
+m4_version_prereq([2.70], [],
+[if test $ac_cv_header_sys_mkdev_h = no; then
+   AC_CHECK_HEADER(sys/sysmacros.h, [AC_DEFINE(MAJOR_IN_SYSMACROS, 1,
+      [Define to 1 if `major', `minor', and `makedev' are declared in <sys/sysmacros.h>.])])
+fi])
+
 # Checks for typedefs, structures, and compiler characteristics.
 # ----------------------------------------------------------------------
 AC_HEADER_STDBOOL

--- a/linux/LinuxProcessList.c
+++ b/linux/LinuxProcessList.c
@@ -26,6 +26,12 @@ in the source distribution for its full text.
 #include <assert.h>
 #include <sys/types.h>
 #include <fcntl.h>
+#ifdef MAJOR_IN_MKDEV
+#include <sys/mkdev.h>
+#elif defined(MAJOR_IN_SYSMACROS) || \
+   (defined(HAVE_SYS_SYSMACROS_H) && HAVE_SYS_SYSMACROS_H)
+#include <sys/sysmacros.h>
+#endif
 
 #ifdef HAVE_DELAYACCT
 #include <netlink/attr.h>

